src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java

```diff
@@ -63,37 +78,158 @@ public void setHbDataSenderService(HBDataSenderService hbDataSenderService)
+        }
+    }
+
+    /**
+     * License load failure resets the heartbeat back to the default state
+     */
+    public synchronized void onLicenseFail()
+    {
+        boolean newEnabled = isEnabledByDefault();
+
+        if (newEnabled != this.enabled)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("State change of heartbeat");
+            }
+            this.enabled = newEnabled;
+            restartAllCollectorSchedules();
+        }
+    }
+
+    private void restartAllCollectorSchedules()
+    {
+        try
+        {
+            for(HBBaseDataCollector collector : collectors)
+            {
+                scheduleCollector(collector);
+            }
+        }
+        catch (Exception e)
+        {
+            logger.error("Unable to schedule heart beat", e);
+        }
+    }
+
+    /**
+     * Start or stop the hertbeat job depending on whether the heartbeat is enabled or not
+     * @throws SchedulerException
+     */
+    private synchronized void scheduleCollector(HBBaseDataCollector collector) throws SchedulerException
+    {
+        // Schedule the heart beat to run regularly
+        final String jobName = "heartbeat-" + collector.getCollectorId();
+        final String triggerName = jobName + "-Trigger";
+        if(this.enabled)
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("heartbeat job scheduled");
+            }
+            final JobDetail jobDetail = new JobDetail(jobName, Scheduler.DEFAULT_GROUP, HeartBeatJob.class);
+            jobDetail.getJobDataMap().put("collector", collector);
+            jobDetail.getJobDataMap().put("hbDataSenderService", hbDataSenderService);
+            jobDetail.getJobDataMap().put("jobLockService", jobLockService);
+
+            // Ensure the job wasn't already scheduled in an earlier retry of this transaction
+            scheduler.unscheduleJob(triggerName, Scheduler.DEFAULT_GROUP);
+
+            String cronExpression = collector.getCronExpression();
+            CronTrigger cronTrigger = null;
             try
             {
-                hbDataSenderService.sendData(data);
+                cronTrigger = new CronTrigger(triggerName , Scheduler.DEFAULT_GROUP, cronExpression);
+            }
+            catch (ParseException e)
+            {
+                if (logger.isDebugEnabled())
+                {
+                    logger.debug("Skipping scheduling because of invalid cron expression: " + cronExpression);
+                }
             }
-            catch (Exception e)
+            if(cronTrigger != null)
+            {
+                scheduler.scheduleJob(jobDetail, cronTrigger);
+            }
+        }
+        else
+        {
+            if (logger.isDebugEnabled())
             {
-                // Log exception
-                logger.error(e);
+                logger.debug("heartbeat job unscheduled");
```

Comment:
I think our logger messages should be updated to reflect which collector is being scheduled/unscheduled. Like this we will just see "heartbeat job unscheduled" multiple times in the log